File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/logging/AbstractLogCollector.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.logging;

import org.kaaproject.kaa.client.channel.KaaChannelManager;
import org.kaaproject.kaa.client.channel.LogTransport;
import org.kaaproject.kaa.client.channel.TransportConnectionInfo;
import org.kaaproject.kaa.client.channel.failover.FailoverManager;
import org.kaaproject.kaa.client.channel.failover.FailoverStatus;
import org.kaaproject.kaa.client.context.ExecutorContext;
import org.kaaproject.kaa.client.logging.future.RecordFuture;
import org.kaaproject.kaa.client.logging.memory.MemLogStorage;
import org.kaaproject.kaa.common.TransportType;
import org.kaaproject.kaa.common.endpoint.gen.LogDeliveryErrorCode;
import org.kaaproject.kaa.common.endpoint.gen.LogDeliveryStatus;
import org.kaaproject.kaa.common.endpoint.gen.LogEntry;
import org.kaaproject.kaa.common.endpoint.gen.LogSyncRequest;
import org.kaaproject.kaa.common.endpoint.gen.LogSyncResponse;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponseResultType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

/**
 * Reference implementation of @see LogCollector.
 *
 * @author Andrew Shvayka
 */
public abstract class AbstractLogCollector implements LogCollector, LogProcessor {
  private static final Logger LOG = LoggerFactory.getLogger(AbstractLogCollector.class);

  protected final ExecutorContext executorContext;
  protected final Map<Integer, List<RecordFuture>> deliveryFuturesMap = new HashMap<>();
  protected final Map<Integer, BucketInfo> bucketInfoMap = new ConcurrentHashMap<>();
  private final LogTransport transport;
  private final ConcurrentHashMap<Integer, Future<?>> timeouts = new ConcurrentHashMap<>();
  private final KaaChannelManager channelManager;
  private final FailoverManager failoverManager;
  private final LogFailoverCommand controller;
  private final Object uploadCheckLock = new Object();
  protected LogStorage storage;
  private LogUploadStrategy strategy;
  private LogDeliveryListener logDeliveryListener;
  private boolean uploadCheckInProgress = false;

  /**
   * All-args constructor.
   */
  public AbstractLogCollector(LogTransport transport, ExecutorContext executorContext,
                              KaaChannelManager channelManager, FailoverManager failoverManager) {
    this.strategy = new DefaultLogUploadStrategy();
    this.storage = new MemLogStorage();
    this.controller = new DefaultLogUploadController();
    this.channelManager = channelManager;
    this.transport = transport;
    this.executorContext = executorContext;
    this.failoverManager = failoverManager;
  }

  @Override
  public void setStrategy(LogUploadStrategy strategy) {
    if (strategy == null) {
      throw new IllegalArgumentException("Strategy is null!");
    }
    this.strategy = strategy;
    LOG.info("New log upload strategy was set: {}", strategy);
  }

  @Override
  public void setStorage(LogStorage storage) {
    if (storage == null) {
      throw new IllegalArgumentException("Storage is null!");
    }
    this.storage = storage;
    LOG.info("New log storage was set {}", storage);
  }

  @Override
  public void fillSyncRequest(LogSyncRequest request) {
    if (!isUploadAllowed()) {
      return;
    }

    LogBucket bucket = storage.getNextBucket();

    if (bucket == null || bucket.getRecords().isEmpty()) {
      LOG.trace("No logs to send");
      return;
    }

    List<LogRecord> recordList = bucket.getRecords();

    LOG.trace("Sending {} log records", recordList.size());

    List<LogEntry> logs = new LinkedList<>();
    for (LogRecord record : recordList) {
      logs.add(new LogEntry(ByteBuffer.wrap(record.getData())));
    }

    request.setRequestId(bucket.getBucketId());
    request.setLogEntries(logs);

    final int bucketId = bucket.getBucketId();
    Future<?> timeoutFuture = executorContext.getScheduledExecutor().schedule(new Runnable() {
      @Override
      public void run() {
        if (!Thread.currentThread().isInterrupted()) {
          checkDeliveryTimeout(bucketId);
        } else {
          LOG.debug("Timeout check worker for log bucket: {} was interrupted", bucketId);
        }
      }
    }, strategy.getTimeout(), TimeUnit.SECONDS);

    LOG.info("Adding following bucket id [{}] for timeout tracking", bucket.getBucketId());
    timeouts.put(bucket.getBucketId(), timeoutFuture);
  }

  @Override
  public void onLogResponse(LogSyncResponse logSyncResponse) throws IOException {
    if (logSyncResponse.getDeliveryStatuses() != null) {
      boolean isAlreadyScheduled = false;
      for (LogDeliveryStatus response : logSyncResponse.getDeliveryStatuses()) {
        final int requestId = response.getRequestId();
        final BucketInfo bucketInfo = bucketInfoMap.get(requestId);
        final long arriveTime = System.currentTimeMillis();
        if (bucketInfo != null) {
          bucketInfoMap.remove(requestId);
          if (response.getResult() == SyncResponseResultType.SUCCESS) {
            storage.removeBucket(response.getRequestId());

            if (logDeliveryListener != null) {
              executorContext.getCallbackExecutor().execute(new Runnable() {
                @Override
                public void run() {
                  logDeliveryListener.onLogDeliverySuccess(bucketInfo);
                }
              });
            }

            executorContext.getCallbackExecutor().execute(new Runnable() {
              @Override
              public void run() {
                notifyDeliveryFuturesOnSuccess(bucketInfo, arriveTime);
              }
            });
          } else {
            storage.rollbackBucket(response.getRequestId());

            final LogDeliveryErrorCode errorCode = response.getErrorCode();
            final LogFailoverCommand controller = this.controller;

            executorContext.getCallbackExecutor().execute(new Runnable() {
              @Override
              public void run() {
                strategy.onFailure(controller, errorCode);
              }
            });

            if (logDeliveryListener != null) {
              executorContext.getCallbackExecutor().execute(new Runnable() {
                @Override
                public void run() {
                  logDeliveryListener.onLogDeliveryFailure(bucketInfo);
                }
              });
            }

            isAlreadyScheduled = true;
          }
        } else {
          LOG.warn("BucketInfo is null");
        }
        LOG.info("Removing bucket id from timeouts: {}", response.getRequestId());
        Future<?> timeoutFuture = timeouts.remove(response.getRequestId());
        if (timeoutFuture != null) {
          timeoutFuture.cancel(true);
        } else {
          LOG.warn("TimeoutFuture is null and cannot be canceled");
        }
      }

      if (!isAlreadyScheduled) {
        processUploadDecision(strategy.isUploadNeeded(storage.getStatus()));
      }
    }
  }

  @Override
  public void stop() {
    LOG.debug("Closing storage");
    storage.close();
    LOG.debug("Clearing timeouts map");
    for (Future<?> timeoutFuture : timeouts.values()) {
      timeoutFuture.cancel(true);
    }
    timeouts.clear();
  }

  private void processUploadDecision(LogUploadStrategyDecision decision) {
    switch (decision) {
      case UPLOAD:
        if (isUploadAllowed()) {
          LOG.debug("Going to upload logs");
          transport.sync();
        }
        break;
      case NOOP:
        if (strategy.getUploadCheckPeriod() > 0 && storage.getStatus().getRecordCount() > 0) {
          scheduleUploadCheck();
        }
        break;
      default:
        break;
    }
  }

  protected void scheduleUploadCheck() {
    LOG.trace("Attempt to execute upload check: {}", uploadCheckInProgress);
    synchronized (uploadCheckLock) {
      if (!uploadCheckInProgress) {
        LOG.trace("Scheduling upload check with timeout: {}", strategy.getUploadCheckPeriod());
        uploadCheckInProgress = true;
        executorContext.getScheduledExecutor().schedule(new Runnable() {
          @Override
          public void run() {
            synchronized (uploadCheckLock) {
              uploadCheckInProgress = false;
            }
            uploadIfNeeded();
          }
        }, strategy.getUploadCheckPeriod(), TimeUnit.SECONDS);
      } else {
        LOG.trace("Upload check is already scheduled!");
      }
    }
  }

  private void checkDeliveryTimeout(final int bucketId) {
    LOG.debug("Checking for a delivery timeout of the bucket with id: [{}] ", bucketId);
    Future<?> timeoutFuture = timeouts.remove(bucketId);

    if (timeoutFuture != null) {
      LOG.info("Log delivery timeout detected for the bucket with id: [{}]", bucketId);

      storage.rollbackBucket(bucketId);

      final LogFailoverCommand controller = this.controller;
      executorContext.getCallbackExecutor().execute(new Runnable() {
        @Override
        public void run() {
          strategy.onTimeout(controller);
        }
      });

      if (logDeliveryListener != null) {
        executorContext.getCallbackExecutor().execute(new Runnable() {
          @Override
          public void run() {
            logDeliveryListener.onLogDeliveryTimeout(bucketInfoMap.get(bucketId));
          }
        });
      }

      timeoutFuture.cancel(true);
    } else {
      LOG.trace("No log delivery timeout for the bucket with id [{}] was detected", bucketId);
    }
  }

  private boolean isUploadAllowed() {
    if (timeouts.size() >= strategy.getMaxParallelUploads()) {
      LOG.debug("Ignore log upload: too much pending requests {}, max allowed {}",
              timeouts.size(), strategy.getMaxParallelUploads());
      return false;
    }
    return true;
  }

  protected void uploadIfNeeded() {
    processUploadDecision(strategy.isUploadNeeded(storage.getStatus()));
  }

  @Override
  public void setLogDeliveryListener(LogDeliveryListener logDeliveryListener) {
    this.logDeliveryListener = logDeliveryListener;
  }

  protected void addDeliveryFuture(BucketInfo info, RecordFuture future) {
    synchronized (deliveryFuturesMap) {
      List<RecordFuture> deliveryFutures = deliveryFuturesMap.get(info.getBucketId());
      if (deliveryFutures == null) {
        deliveryFutures = new LinkedList<RecordFuture>();
        deliveryFuturesMap.put(info.getBucketId(), deliveryFutures);
      }

      deliveryFutures.add(future);
    }
  }

  protected void notifyDeliveryFuturesOnSuccess(BucketInfo info, Long arriveTime) {
    synchronized (deliveryFuturesMap) {
      List<RecordFuture> deliveryFutures = deliveryFuturesMap.get(info.getBucketId());
      if (deliveryFutures != null) {
        for (RecordFuture future : deliveryFutures) {
          RecordInfo recordInfo = new RecordInfo(info);
          future.setValue(recordInfo, arriveTime);
        }

        deliveryFuturesMap.remove(info.getBucketId());
      }
    }
  }

  private class DefaultLogUploadController implements LogFailoverCommand {
    @Override
    public void switchAccessPoint() {
      TransportConnectionInfo server = channelManager.getActiveServer(TransportType.LOGGING);
      if (server != null) {
        failoverManager.onServerFailed(server, FailoverStatus.OPERATION_SERVERS_NA);
      } else {
        LOG.warn("Failed to switch Operation server. No channel is used for logging transport");
      }
    }

    @Override
    public void retryLogUpload() {
      uploadIfNeeded();
    }

    @Override
    public void retryLogUpload(int delay) {
      executorContext.getScheduledExecutor().schedule(new Runnable() {
        @Override
        public void run() {
          uploadIfNeeded();
        }
      }, delay, TimeUnit.SECONDS);
    }
  }
}

File: client/client-multi/client-objective-c/Kaa/logging/AbstractLogCollector.m
/*
 * Copyright 2014-2015 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "AbstractLogCollector.h"
#import "DefaultLogUploadStrategy.h"
#import "MemLogStorage.h"
#import "LogFailoverCommand.h"
#import "KaaChannelManager.h"
#import "FailoverManager.h"
#import "KaaLogging.h"
#import "LogCollector.h"

#define TAG @"AbstractLogCollector >>>"

@interface AbstractLogCollector () <LogFailoverCommand>

@property (nonatomic,strong) id<LogUploadStrategy> strategy;
@property (nonatomic,strong) id<KaaChannelManager> channelManager;
@property (nonatomic,strong) id<LogTransport> transport;
@property (nonatomic,strong) id<FailoverManager> failoverManager;
@property (nonatomic,strong) NSMutableDictionary *timeouts;
@property (nonatomic,strong) NSLock *timeoutsLock;
@property (atomic) BOOL uploadCheckInProgress;
@property (nonatomic,strong) NSLock *uploadCheckLock;
@property (nonatomic,strong) NSObject *uploadCheckGuard;   //variable to sync
@property (nonatomic,weak) id<LogDeliveryDelegate> logDeliveryDelegate;

- (void)checkDeliveryTimeout:(int32_t)bucketId;
- (void)processUploadDecision:(LogUploadStrategyDecision)decision;

@end

@interface TimeoutOperation : NSOperation

@property (nonatomic) int64_t timeout;
@property (nonatomic,weak) AbstractLogCollector *logCollector;
@property (nonatomic,strong) LogBlock *timeoutGroup;

- (instancetype)initWithLogCollector:(AbstractLogCollector *)logCollector timeout:(int64_t)timeout andGroup:(LogBlock *)group;

@end

@implementation AbstractLogCollector

- (instancetype)initWith:(id<LogTransport>)transport
         executorContext:(id<ExecutorContext>)executorContext
          channelManager:(id<KaaChannelManager>)channelManager
         failoverManager:(id<FailoverManager>)failoverManager {
    self = [super init];
    if (self) {
        self.strategy = [[DefaultLogUploadStrategy alloc] initWithDefaults];
        self.storage = [[MemLogStorage alloc] initWithBucketSize:[self.strategy getBatchSize] bucketRecordCount:[self.strategy getBatchCount]];
        self.channelManager = channelManager;
        self.transport = transport;
        _executorContext = executorContext;
        self.failoverManager = failoverManager;
        self.timeouts = [NSMutableDictionary dictionary];
        self.timeoutsLock = [[NSLock alloc] init];
        self.uploadCheckInProgress = NO;
        self.uploadCheckLock = [[NSLock alloc] init];
        self.uploadCheckGuard = [[NSObject alloc] init];
        self.logDeliveryDelegate = nil;
        self.bucketInfoDictionary = [NSMutableDictionary dictionary];
        self.bucketRunnerDictionary = [NSMutableDictionary dictionary];
    }
    return self;
}

- (void)setLogDeliveryDelegate:(id<LogDeliveryDelegate>)logDeliveryDelegate {
    _logDeliveryDelegate = logDeliveryDelegate;
}

- (void)setStrategy:(id<LogUploadStrategy>)strategy {
    if (!strategy) {
        [NSException raise:NSInvalidArgumentException format:@"%@ Strategy is nil!", TAG];
    }
    _strategy = strategy;
    DDLogInfo(@"%@ New log upload strategy was set: %@", TAG, strategy);
}

- (void)setStorage:(id<LogStorage>)storage {
    if (!storage) {
        [NSException raise:NSInvalidArgumentException format:@"%@ Storage is nil!", TAG];
    }
    _storage = storage;
    DDLogInfo(@"%@ New log storage was set: %@", TAG, storage);
}

- (void)fillSyncRequest:(LogSyncRequest *)request {
    LogBlock *group = nil;
    if ([[self.storage getStatus] getRecordCount] == 0) {
        DDLogDebug(@"%@ Log storage is empty", TAG);
        return;
    }
    if (![self isUploadAllowed]) {
        return;
    }
    group = [self.storage getRecordBlock];
    if (group) {
        NSArray *recordList = group.logRecords;
        if ([recordList count] > 0) {
            DDLogVerbose(@"%@ Sending %li log records", TAG, (long)[recordList count]);
            NSMutableArray *logs = [NSMutableArray array];
            for (LogRecord *record in recordList) {
                LogEntry *logEntry = [[LogEntry alloc] init];
                logEntry.data = [NSData dataWithData:record.data];
                [logs addObject:logEntry];
            }
            request.requestId = group.blockId;
            request.logEntries = [KAAUnion unionWithBranch:KAA_UNION_ARRAY_LOG_ENTRY_OR_NULL_BRANCH_0 andData:logs];
            
            DDLogInfo(@"%@ Adding following bucket id [%i] for timeout tracking", TAG, group.blockId);
            NSOperation *timeoutOperation = [[TimeoutOperation alloc] initWithLogCollector:self
                                                                                   timeout:[self.strategy getTimeout]
                                                                                  andGroup:group];

            [self.timeoutsLock lock];
            [self.timeouts setObject:timeoutOperation forKey:[NSNumber numberWithLong:group.blockId]];
            [self.timeoutsLock unlock];
        }
    } else {
        DDLogWarn(@"%@ Log group is nil: log group size is too small", TAG);
    }
}

- (void)onLogResponse:(LogSyncResponse *)response {
    @synchronized (self) {
        if (response.deliveryStatuses && response.deliveryStatuses.branch == KAA_UNION_ARRAY_LOG_DELIVERY_STATUS_OR_NULL_BRANCH_0) {
            BOOL isAlreadyScheduled = NO;
            NSArray *deliveryStatuses = response.deliveryStatuses.data;
            __weak typeof(self) weakSelf = self;
            for (LogDeliveryStatus *status in deliveryStatuses) {
                if (status.result == SYNC_RESPONSE_RESULT_TYPE_SUCCESS) {
                    [self.storage removeRecordBlock:status.requestId];
                    BucketInfo *bucketInfo = [self.bucketInfoDictionary objectForKey:[NSNumber numberWithInt:status.requestId]];
                    BucketRunner *currentRunner = [self.bucketRunnerDictionary objectForKey:[NSNumber numberWithInt:status.requestId]];
                    [currentRunner setValue:bucketInfo];
                    if (self.logDeliveryDelegate) {
                        [[self.executorContext getCallbackExecutor] addOperationWithBlock:^{
                            [weakSelf.logDeliveryDelegate onLogDeliverySuccess:bucketInfo];
                        }];
                    }
                    [self.bucketInfoDictionary removeObjectForKey:[NSNumber numberWithInt:status.requestId]];
                } else {
                    [self.storage notifyUploadFailed:status.requestId];
                    [[self.executorContext getCallbackExecutor] addOperationWithBlock:^{
                        [weakSelf.strategy onFailure:weakSelf errorCode:[((NSNumber *)status.errorCode.data) intValue]];
                        if (weakSelf.logDeliveryDelegate) {
                            [weakSelf.logDeliveryDelegate onLogDeliveryFailure:[weakSelf.bucketInfoDictionary objectForKey:[NSNumber numberWithInt:status.requestId]]];
                        }
                    }];
                    isAlreadyScheduled = YES;
                }
                DDLogInfo(@"%@ Removing bucket id from timeouts: %i", TAG, status.requestId);
                [self.timeoutsLock lock];
                NSNumber *key = [NSNumber numberWithLong:status.requestId];
                NSOperation *timeout = [self.timeouts objectForKey:key];
                if (timeout) {
                    [self.timeouts removeObjectForKey:key];
                    [timeout cancel];
                }
                [self.timeoutsLock unlock];
            }
            if (!isAlreadyScheduled) {
                [self processUploadDecision:[self.strategy isUploadNeeded:[self.storage getStatus]]];
            }
        }
    }
}

- (void)stop {
    DDLogDebug(@"%@ Closing storage", TAG);
    [self.storage close];
    DDLogDebug(@"%@ Clearing timeouts map", TAG);
    for (NSOperation *timeout in self.timeouts.allValues) {
        [timeout cancel];
    }
    [self.timeouts removeAllObjects];
}

- (void)processUploadDecision:(LogUploadStrategyDecision)decision {
    switch (decision) {
        case LOG_UPLOAD_STRATEGY_DECISION_UPLOAD:
            if ([self isUploadAllowed]) {
                [self.transport sync];
            }            break;
        case LOG_UPLOAD_STRATEGY_DECISION_NOOP:
            if ([self.strategy getUploadCheckPeriod] > 0 && [[self.storage getStatus] getRecordCount] > 0) {
                [self scheduleUploadCheck];
            }
            break;
        default:
            break;
    }
}

- (void)scheduleUploadCheck {
    DDLogVerbose(@"%@ Attempt to execute upload check: %i", TAG, self.uploadCheckInProgress);
    @synchronized(self.uploadCheckGuard) {
        if (!self.uploadCheckInProgress) {
            DDLogVerbose(@"%@ Scheduling upload check with timeout: %li", TAG, (long)[self.strategy getUploadCheckPeriod]);
            self.uploadCheckInProgress = YES;
            __weak typeof(self)weakSelf = self;
            dispatch_time_t delay = dispatch_time(DISPATCH_TIME_NOW, (int64_t)([self.strategy getUploadCheckPeriod] * NSEC_PER_SEC));
            dispatch_after(delay, [self.executorContext getSheduledExecutor], ^{
                
                @synchronized(self.uploadCheckGuard) {
                    weakSelf.uploadCheckInProgress = NO;
                }
                
                [weakSelf uploadIfNeeded];
            });
        } else {
            DDLogVerbose(@"%@ Upload check is already scheduled!", TAG);
        }
    }
}

- (void)checkDeliveryTimeout:(int32_t)bucketId {
    DDLogDebug(@"%@ Checking for a delivery timeout of the bucket with id: [%i]", TAG, bucketId);
    [self.timeoutsLock lock];
    NSOperation *timeout = [self.timeouts objectForKey:[NSNumber numberWithLong:bucketId]];
    if (timeout) {
        [self.timeouts removeObjectForKey:[NSNumber numberWithLong:bucketId]];
    }
    [self.timeoutsLock unlock];
    
    if (timeout) {
        DDLogInfo(@"%@ Log delivery timeout detected for the bucket with id: [%i]", TAG, bucketId);
        [self.storage notifyUploadFailed:bucketId];
        __weak typeof(self)weakSelf = self;
        [[self.executorContext getCallbackExecutor] addOperationWithBlock:^{
            [weakSelf.strategy onTimeout:weakSelf];
            if (weakSelf.logDeliveryDelegate) {
                [weakSelf.logDeliveryDelegate onLogDeliveryTimeout:[self.bucketInfoDictionary objectForKey:[NSNumber numberWithInt:bucketId]]];
            }
        }];
        [timeout cancel];
    } else {
        DDLogVerbose(@"%@ No log delivery timeout for the bucket with id [%li] was detected", TAG, (long)bucketId);
    }
}

- (void)uploadIfNeeded {
    [self processUploadDecision:[self.strategy isUploadNeeded:[self.storage getStatus]]];
}

- (void)switchAccessPoint {
    id<TransportConnectionInfo> server = [self.channelManager getActiveServer:TRANSPORT_TYPE_LOGGING];
    if (server) {
        [self.failoverManager onServerFailed:server];
    } else {
        DDLogWarn(@"%@ Failed to switch Operation server. No channel is used for logging transport", TAG);
    }
}

- (void)retryLogUpload {
    [self uploadIfNeeded];
}

- (void)retryLogUpload:(int32_t)delay {
    __weak typeof(self)weakSelf = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delay * NSEC_PER_SEC)), [self.executorContext getSheduledExecutor], ^{
        [weakSelf uploadIfNeeded];
    });
}

- (BOOL)isUploadAllowed {
    if (self.timeouts.count >= [self.strategy getMaxParallelUploads]) {
        DDLogDebug(@"%@ Ignore log upload: too much pending requests. Max allowed: %lld", TAG, [self.strategy getMaxParallelUploads]);
        return NO;
    }
    return YES;
}

@end

@implementation TimeoutOperation

- (instancetype)initWithLogCollector:(AbstractLogCollector *)logCollector timeout:(int64_t)timeout andGroup:(LogBlock *)group {
    self = [super init];
    if (self) {
        self.logCollector = logCollector;
        self.timeout = timeout;
        self.timeoutGroup = group;
    }
    return self;
}

- (void)main {
    if (self.isFinished && self.isCancelled) {
        DDLogDebug(@"%@ Timeout check worker for block: %i was interrupted before start", TAG, [self.timeoutGroup blockId]);
        return;
    }
    
    [NSThread sleepForTimeInterval:self.timeout];
    
    if (!self.isFinished && !self.isCancelled) {
        [self.logCollector checkDeliveryTimeout:[self.timeoutGroup blockId]];
        
    } else {
        DDLogDebug(@"%@ Timeout check worker for block: %i was interrupted after delay", TAG, [self.timeoutGroup blockId]);
    }
}

@end
